service/history/queue/processing_queue_test.go

```diff
@@ -110,6 +110,140 @@ func (s *processingQueueSuite) TestAddTasks() {
 	s.Equal(newReadLevel, queue.state.readLevel)
 }
 
+func (s *processingQueueSuite) TestGetTask_TaskNotFound_Error() {
+	ackLevel := testKey{ID: 1}
+	maxLevel := testKey{ID: 10}
+
+	taskKeys := []task.Key{
+		testKey{ID: 2},
+		testKey{ID: 3},
+		testKey{ID: 5},
+		testKey{ID: 9},
+	}
+	tasks := make(map[task.Key]task.Task)
+	for _, key := range taskKeys {
+		mockTask := task.NewMockTask(s.controller)
+		mockTask.EXPECT().GetDomainID().Return("some random domainID").AnyTimes()
+		mockTask.EXPECT().GetWorkflowID().Return("some random workflowID").AnyTimes()
+		mockTask.EXPECT().GetRunID().Return("some random runID").AnyTimes()
+		mockTask.EXPECT().GetTaskType().Return(0).AnyTimes()
+		tasks[key] = mockTask
+	}
+
+	queue := s.newTestProcessingQueue(
+		0,
+		ackLevel,
+		ackLevel,
+		maxLevel,
+		NewDomainFilter(nil, true),
+		make(map[task.Key]task.Task),
+	)
+
+	newReadLevel := testKey{ID: 10}
+	queue.AddTasks(tasks, newReadLevel)
+	s.Len(queue.outstandingTasks, len(taskKeys))
+
+	task, err := queue.GetTask(testKey{ID: 100})
+	s.Nil(task)
+	s.Error(err)
+}
+
+func (s *processingQueueSuite) TestGetTask_Success() {
+	ackLevel := testKey{ID: 1}
+	maxLevel := testKey{ID: 10}
+
+	taskKeys := []task.Key{
+		testKey{ID: 2},
+		testKey{ID: 3},
+		testKey{ID: 5},
+		testKey{ID: 9},
+	}
+	tasks := make(map[task.Key]task.Task)
+	for _, key := range taskKeys {
+		mockTask := task.NewMockTask(s.controller)
+		mockTask.EXPECT().GetDomainID().Return("some random domainID").AnyTimes()
+		mockTask.EXPECT().GetWorkflowID().Return("some random workflowID").AnyTimes()
+		mockTask.EXPECT().GetRunID().Return("some random runID").AnyTimes()
+		mockTask.EXPECT().GetTaskType().Return(0).AnyTimes()
+		tasks[key] = mockTask
+	}
+
+	queue := s.newTestProcessingQueue(
+		0,
+		ackLevel,
+		ackLevel,
+		maxLevel,
+		NewDomainFilter(nil, true),
+		make(map[task.Key]task.Task),
+	)
+
+	newReadLevel := testKey{ID: 10}
```

Comment:
What does level mean in this context? I have seen it used with ackLevel, readLevel and readMaxLevel.

Are those offsets where sender and receiver send to each other. Is readLevel the offset receiver uses to read tasks starting from the given id and ackLevel the offset used by the sender to understand until which task id the tasks were transferred successfully?

Reply:
Ack level is the max task id that has been processed. 
Read level is the max task id loaded into the queue(in memory)
Max read level is the max task id that can be read from DB. 

Those offsets are stored in x cross queue only.